src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,204 @@
+        Node<T> node = new Node<>(tail, value, null);
+        if (head == null) {
+            head = node;
+            tail = head;
+        } else {
+            tail.next = node;
+            tail = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        isValidIndexForAdd(index);
+        if (isEmpty()) {
+            add(value);
+        } else if (index == 0) {
+            addToHead(value);
+        } else {
+            Node<T> previous = findNodeByIndex(index);
+            if (previous == null) {
+                addToTail(value);
+            } else {
+                insert(value, previous);
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isValidIndexForSetAndGet(index);
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isValidIndexForSetAndGet(index);
+        Node<T> foundNode = findNodeByIndex(index);
+        T oldValue = foundNode.value;
+        foundNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> deletedElement = findNodeByIndex(index);
+        if (deletedElement == null) {
+            throw new IndexOutOfBoundsException("wrong index: " + index + ", for empty list");
+        }
+        unlink(deletedElement);
+        return deletedElement.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> deletedElement = findNodeByElement(object);
+        boolean isDeleted = unlink(deletedElement);
+        return isDeleted;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void insert(T value, Node<T> previous) {
+        Node<T> newNode = new Node(previous.prev, value, previous);
+        previous.prev.next = newNode;
+        previous.prev = newNode;
+        size++;
+    }
+
+    private void addToHead(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        head.prev = newNode;
+        head = newNode;
+        size++;
+    }
+
+    private void addToTail(T value) {
+        Node<T> newNode = new Node<>(tail, value,null);
+        tail.next = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private void isValidIndexForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("invalid index: " + index);
+        }
+    }
+
+    private void isValidIndexForSetAndGet(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("invalid index: " + index);
+        }
+    }
```

Comment:
one of these methods (for add) is redundant and could be removed. Another one should be renamed. For example, you could use 'checkIndex' or 'validateIndex'